src/resolve/resolved-dns-browse-services.c

```diff
@@ -0,0 +1,788 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include "af-list.h"
+#include "event-util.h"
+#include "dns-domain.h"
+#include "json-util.h"
+#include "random-util.h"
+#include "resolved-dns-browse-services.h"
+#include "resolved-dns-cache.h"
+#include "resolved-dns-question.h"
+#include "resolved-dns-query.h"
+#include "resolved-dns-rr.h"
+#include "resolved-dns-scope.h"
+#include "resolved-manager.h"
+#include "resolved-varlink.h"
+#include "string-table.h"
+
+typedef enum BrowseServiceUpdateEvent {
+        BROWSE_SERVICE_UPDATE_ADDED,
+        BROWSE_SERVICE_UPDATE_REMOVED,
+        _BROWSE_SERVICE_UPDATE_MAX,
+        _BROWSE_SERVICE_UPDATE_INVALID = -EINVAL,
+} BrowseServiceUpdateEvent;
+
+static const char * const browse_service_update_event_table[_BROWSE_SERVICE_UPDATE_MAX] = {
+        [BROWSE_SERVICE_UPDATE_ADDED]   = "added",
+        [BROWSE_SERVICE_UPDATE_REMOVED] = "removed",
+};
+
+DEFINE_PRIVATE_STRING_TABLE_LOOKUP_TO_STRING(browse_service_update_event, BrowseServiceUpdateEvent);
+
+/* RFC6762 5.2
+ * The intervals between successive queries MUST increase by at least a
+ * factor of two. When the interval between queries reaches or exceeds
+ * 60 minutes, perform subsequent queries at a steady-state rate of one
+ * query per hour */
+usec_t mdns_calculate_next_query_delay(usec_t current_delay) {
+        assert(current_delay > 0);
+        assert(current_delay <= 3600);
+
+        return current_delay < 2048 ? current_delay * 2 : 3600;
```

Comment:
hmm, so if we are at 2047, then this will result in 4094 which is > 3600, and would then hit the assert() on next iteration? 

this is really weird hence?